fix: first round fee stake

mtsalenc opened the pull request:
In the current version, the GTCR requires that the submitter pay the fee stake to make a request and to challenge a request. 

This is wrong since there is no crowdfunding taking place in the first round. It has the effect of increasing the cost to make a submission or challenge a request and requires developers to also account for it when creating displays.

This PR fixes that by only requiring the base deposit and arbitration fees.

Note on field renaming:
The `paidFees` field name was a bit misleading since it did not include just arbitration and appeal fees, but also the base deposit.
Since I needed to create a new field specific for arbitration fees I went ahead and renamed the variable to avoid confusion.

unknownunknown1 commented:
L68: "Tracks the sum for each Party" - I think it's better to add the word "paid" here -> "Tracks the sum paid for each Party" so it'd be more consistent with the rest, like L72 for instance.
L369: I understand what the logic is but in practice I think this line is pretty flawed. Each time someone funds an appeal this variable will be increased for the amount of appealCost regardless of the funded sum. You can even pay 0 eth many times and greatly inflate this variable. Maybe this line should be removed since it's not used in appeal rounds anyway.
L419: This condition seems unnecessary since we enter this block only if both sides paid their fees, unless 0 eth is enough to mark hasPaid as 'true' which I'm not think is possible in practice.
L409-L436: I'm not sure I understand the importance of this section.  Don't we already reimburse the sides proportionally to their contributions? 
`round.contributions[_beneficiary][uint(Party.Requester)] * round.feeRewards) / (round.amountPaid[uint(Party.Challenger)] + round.amountPaid[uint(Party.Requester)])` (L440).
So just in case I checked the old and the new approaches. If we have 1 eth base deposit for both sides, and arbitration fees of 2 eth (requester) and 4 eth (challenger) then with the old approach the rewards are 1.5 and 2.5 for requester and challenger respectively. And with the new approach it's 1.66 and 2.34. If the fee doesn't change then the numbers are the same in both cases. I don't see a huge difference there that would justify this new complexity.
L810, L820: If we decide to keep the new variable I think it should be added to getters as well.

clesaege commented on contracts/GeneralizedTCR.sol:
```diff
@@ -403,20 +406,48 @@ contract GeneralizedTCR is IArbitrable, IEvidence {
+                _beneficiary != request.parties[uint(Party.Challenger)]
+            ) return;
+
+            uint side = _beneficiary == request.parties[uint(Party.Requester)] ? 1 : 2;
```
We should avoid magic numbers. We can use Party.X and cast it into an uint only at the last moment.

mtsalenc commented:
> L68: "Tracks the sum for each Party" - I think it's better to add the word "paid" here -> "Tracks the sum paid for each Party" so it'd be more consistent with the rest, like L72 for instance.

Fixed

> L369: I understand what the logic is but in practice I think this line is pretty flawed. Each time someone funds an appeal this variable will be increased for the amount of appealCost regardless of the funded sum. You can even pay 0 eth many times and greatly inflate this variable. Maybe this line should be removed since it's not used in appeal rounds anyway.

You are right this was a real bug. Nice catch, I removed the line entirely.

> L419: This condition seems unnecessary since we enter this block only if both sides paid their fees, unless 0 eth is enough to mark hasPaid as 'true' which I don't think is possible in practice.

Fixed.

> L409-L436: I'm not sure I understand the importance of this section. Don't we already reimburse the sides proportionally to their contributions?
> `round.contributions[_beneficiary][uint(Party.Requester)] * round.feeRewards) / (round.amountPaid[uint(Party.Challenger)] + round.amountPaid[uint(Party.Requester)])` (L440).
> So just in case I checked the old and the new approaches. If we have 1 eth base deposit for both sides, and arbitration fees of 2 eth (requester) and 4 eth (challenger) then with the old approach the rewards are 1.5 and 2.5 for requester and challenger respectively. And with the new approach it's 1.66 and 2.34. If the fee doesn't change then the numbers are the same in both cases. I don't see a huge difference there that would justify this new complexity.

The key point behind this is that with the previous approach, in the case of refusing to rule, parties would be reimbursed proportionally to the total amount contributed `baseDeposit`s + arbitration fees. This is a problem if the `requestBaseDeposit`and `challengeBaseDeposit` are significantly different because when parties are reimbursed they may pay or receive more or less than they should've. Instead, we want them to have the `baseDeposit` reimbursed in full and only charge half of the arbitration fees.  As an example consider the below parameters:

`arbitrationCost`: 500
`submissionBaseDeposit`: 50000
`submissionChallengeBaseDeposit`: 1000

Each party should pay only 250 in arbitration fees, however we find that: 

Requester paid  ~`485` instead of `250`
Challenger paid ~`15` instead of `250`

Which is unfair, however with the added complexity each party will always pay 50% of the arbitration fees + their deposit.

I may have made some mistake in my tests but I made a branch `demo/remibursement` to demonstrate. You can check it out and run the test to see what I mean.

> L810, L820: If we decide to keep the new variable I think it should be added to getters as well.

Fixed

unknownunknown1 commented:
I understand. It is indeed more fair, though in practice how common is that for the deposits to be vastly different? Because if the deposits are 2 and 4 the difference in the reward is not that significant to justify the complexity I think.

mtsalenc commented:
One of the cases was the Ethfinex badge, if I recall submission cost around 21 ETH while challenging it cost a lot less. This can happen when we want to give a lot of reward for challenging a case but keep the challenge cost accessible.

For such high stake badges (such as for say submission cost of 10k USD and challenge 1k USD) people would get mad if they realized they over a hundred dollar more than the other side so I think the added complexity is worth it.

clesaege commented:
The behaviour of refusing to rule is not really important and not defined in the spec, so I would keep it simple.

mtsalenc commented:
While a big difference between challenge and submission deposits may not be rare, I agree that the refuse to rule case is rare so I removed the complex code.

unknownunknown1 commented:
Now it seems that `paidArbitrationFees` isn't used anywhere and can be removed. I don't see any other issues with the main file

mtsalenc commented:
The last commit adds a new event `RequestEvidenceGroupID` to replace the old `evidenceGroupIDToRequestID` mapping. Ideally we should not have to create a new event just for this and instead, add the `evidenceGroupID` field to the `RequestSubmitted` event that is already emitted when a user makes a request. Unfortunately the maximum number of indexable parameter in an event is 3.

clesaege commented:
The ruling 0 is handled as expected.
